account_billing/models/account_billing.py

```diff
@@ -0,0 +1,202 @@
+        default=lambda self: self._context.get('active_ids', []),
+        string='Invoices',
+    )
+    invoice_related_count = fields.Integer(
+        string='# of Invoices',
+        compute='_compute_invoice_related_count',
+        help='Count invoice in billing',
+    )
+    state = fields.Selection(
+        [('draft', 'Draft'),
+         ('cancel', 'Cancelled'),
+         ('billed', 'Billed')],
+        string='Status',
+        readonly=True,
+        default='draft',
+        help="""
+            * The 'Draft' status is used when a user create a new billing\n
+            * The 'Billed' status is used when user confirmed billing,
+                billing number is generated\n
+            * The 'Cancelled' status is used when user billing is cancelled
+        """,
+    )
+    narration = fields.Text(
+        string='Notes',
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+        help="Notes",
+    )
+    bill_type = fields.Selection(
+        [('out_invoice', 'Customer Invoice'),
+         ('in_invoice', 'Vendor Bill')],
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+        default=lambda self: self._context.get('bill_type', False)
+    )
+    currency_id = fields.Many2one(
+        'res.currency',
+        string='Currency',
+        required=True,
+        default=lambda self: self._get_currency_id(),
+        help='Currency',
+    )
+
+    @api.model
+    def fields_view_get(self, view_id=None, view_type='form',
+                        toolbar=False, submenu=False):
+        res = super(AccountBilling, self).fields_view_get(
+            view_id=view_id, view_type=view_type,
+            toolbar=toolbar, submenu=submenu)
+        invoices = self.env['account.invoice'].browse(
+            self._context.get('active_ids', []))
+        billing = invoices.mapped('billing_id')
+        if any(inv.state != 'open' for inv in invoices):
+            raise ValidationError(_('The billing cannot be processed '
+                                    'because the invoice is not open!'))
+        if billing:
+            raise ValidationError(_(
+                'Invoice: %s, already billed!') % ', '.join(
+                invoices.filtered(lambda l: l.billing_id).mapped('number')))
+        return res
+
+    @api.onchange('bill_type')
+    def _onchange_bill_type(self):
+        # Set partner_id domain
+        if self.bill_type:
+            return {'domain': {'partner_id': [
+                (MAP_INVOICE_TYPE_PARTNER_TYPE[self.bill_type], '=', True)]}}
+
+    @api.multi
+    def _get_partner_id(self):
+        partner_ids = self.env['account.invoice'].browse(
+            self._context.get('active_ids', [])).mapped('partner_id')
+        if len(partner_ids) > 1:
+            raise ValidationError(
+                _('Please select invoices with same partner'))
+        return partner_ids
+
+    @api.multi
+    def _get_currency_id(self):
```

Comment:
It is good to check for currency here, I think this is when you select invoice in tree view and use action, right?
What about if user select invoices after new billing, do you also check for currency mix?

Reply:
User can't select invoice currency mix. When user selected currency in billing, Invoice are filter by currency